src/common/type/pack.h

```diff
@@ -0,0 +1,500 @@
+/***********************************************************************************************************************************
+Pack Type
+
+The pack type encodes binary data compactly while still maintaining structure and strict typing. The idea is based on Thrift,
+ProtocolBuffers, and Avro, compared here: https://medium.com/better-programming/use-binary-encoding-instead-of-json-dec745ec09b6.
+The pack type has been further optimized to balance between purely in-memory structures and those intended to be passed via a
+protocol or saved in a file.
+
+Integers are stored with base-128 varint encoding which is equivalent to network byte order, i.e., the endianness of the sending and
+receiving host don't matter.
+
+The overall idea is similar to JSON but IDs are used instead of names, typing is more granular, and the representation is far more
+compact. A pack can readily be converted to JSON but the reverse is not as precise due to loose typing in JSON. A pack is a stream
+format, i.e. it is intended to be read in order from beginning to end.
+
+Fields in a pack are identified by IDs. A field ID is stored as a delta from the previous ID, which is very efficient, but means
+that reading from the middle is generally not practical. The size of the gap between field IDs is important -- a gap of 1 never
+incurs extra cost, but depending on the field type larger gaps may require additional bytes to store the field ID delta.
+
+NULLs are not stored in a pack and are therefore not typed. A NULL is essentially just a gap in the field IDs. Fields that are
+frequently NULL are best stored at the end of an object.
+
+A pack is an object by default. Objects can store fields, objects, or arrays. Objects and arrays will be referred to collectively as
+containers. Fields contain data to be stored, e.g. integers, strings, etc.
+
+Here is a simple example of a pack:
+
+PackWrite *write = pckWriteNew(buffer);
+pckWriteU64P(write, 77);
```

Comment:
U64P, BoolP, I32P - these all make me think we're reading pointers. I strongly recommend we drop the "P" on the end since we're reading a "type". And, how do we read a pointer anyway?

Reply:
We have convention to suffix parameterized function macros with P. If there's confusion with the `FUNCTION_LOG_PARAM_P()` macro (and others like it) I vote we rename those macro, .e.g `FUNCTION_LOG_PARAM_PTR()`, which would follow our general convention of using `ptr` in function names when specifically dealing with pointers.

Reply:
ok